src/dataflow/source/kinesis.rs

```diff
@@ -82,78 +83,96 @@ where
-            while let Some(iterator) = &shard_iterator {
-                let get_records_output = match block_on(get_records(&client, &iterator)) {
-                    Ok(output) => {
-                        *shard_iterator = output.next_shard_iterator.clone();
-                        output
-                    }
-                    Err(RusotoError::HttpDispatch(e)) => {
-                        // todo@jldlaughlin: Parse this to determine fatal/retriable?
-                        error!("{}", e);
-                        activator.activate_after(get_reactivation_duration(timer));
-                        return SourceStatus::Alive;
-                    }
-                    Err(RusotoError::Service(GetRecordsError::ExpiredIterator(e))) => {
-                        // todo@jldlaughlin: Will need track source offsets to grab a new iterator.
-                        error!("{}", e);
-                        return SourceStatus::Done;
-                    }
-                    Err(RusotoError::Service(GetRecordsError::ProvisionedThroughputExceeded(
-                        _,
-                    ))) => {
-                        activator.activate_after(get_reactivation_duration(timer));
-                        return SourceStatus::Alive;
-                    }
-                    Err(e) => {
-                        // Fatal service errors:
-                        //  - InvalidArgument
-                        //  - KMSAccessDenied, KMSDisabled, KMSInvalidState, KMSNotFound,
-                        //    KMSOptInRequired, KMSThrottling
-                        //  - ResourceNotFound
-                        //
-                        // Other fatal Rusoto errors:
-                        // - Credentials
-                        // - Validation
-                        // - ParseError
-                        // - Unknown (raw HTTP provided)
-                        // - Blocking
-                        error!("{}", e);
-                        return SourceStatus::Done;
-                    }
-                };
+            // Rotate through the queue of shards. Start with a new
+            // shard on each activation.
+            while let Some(shard_id) = shard_queue.pop_front() {
+                // While the next_shard_iterator is Some(iterator), the shard is open
+                // and could return more data.
+                while let Some(iterator) = shard_to_iterator.get(&shard_id).unwrap_or_else(|| {
+                    error!("Expected to find shard iterator for shard {}, found None. Closing shard...", shard_id);
+                    &None
+                }) {
+                    {
+                        // Pushing back to the shard_queue will allow us to read from the
+                        // shard again.
+                        shard_queue.push_back(shard_id.clone());
+                        let get_records_output = match block_on(get_records(&client, &iterator)) {
```

Comment:
Given get_records is async anyways, does it make sense to add it to a list of futures, loop through the queue of shard iterators, and then once the loop is over, just join? This would potentially also address your hotspot issue too that you mention in a PR?